_posts/2015-10-22-transcending-rest-and-rpc.md

```diff
@@ -0,0 +1,102 @@
+Also, CQRS is amazing in the fact that you could decide to model your system around different bounded contexts, i.e, different CQRS deployables. If your client needs a view that aggregate data from two CQRS units, you have the same problem.
+
+## Demand driven architecture to the rescue
+
+This leads to the point of Demand Driven Architectures. Conceptually the backend is just a single entity and, since it's modeled using REST principles, it's defining what is the shape of the data. However, we're going to have tons of clients with different needs and all of them need to twist themselves in order to get the shape of the data that they want. Would it be great if clients could just send queries to the logical data unit called backend?
+
+If we think in a relational database the concept is similar. We have different tables (rest endpoints, resources, backend services) and we have a query language that allows us to join and shape the data easily. Some of the optimizations of that query will be done by the platform that runs those queries.
+
+## Falcor and JSON Graph
+
+Netflix has created a tool called [Falcor](http://netflix.github.io/falcor) that tries to solve this concrete problem. Clients won't talk with the backend services directly anymore, instead of they will send queries to Falcor service that will do the routing, aggregations, projections and optimizations. This is an image from Falcor's official documentation.
+
+<img src="/assets/img/custom/blog/falcor-network-diagram.png" />
+
+They based that platform in the idea of JSON Graph. Let me explain it briefly. JSON is in essence a tree structure. However, most of our data has relationships so it should be modeled like a graph. Using REST design programmers usually solve that problem through ids, but that makes our live really difficult if you think in problems like cache invalidation. As they say:
+
+> JSON Graph is valid JSON and can be parsed by any JSON parser. However JSON Graph introduces new primitive types to JSON to allow JSON to be used to represent graph information in a simple and consistent way.
+
+They borrow the idea from unix filesystem. That filesystem is a tree, but thanks to symlinks you can emulate the behaviour of a graph. That has massive implications in the way that Falcor optimises calls to backend services. Instead of keeping some values in a denormalised fashion, Falcor keeps references to that normalised value, so, if you ask for a value that has been already retrieved, Falcor will avoid that call. Also that makes cache invalidation business much more easier.
+
+<img src="/assets/img/custom/blog/falcor-services-diagram.png" />
+
+As you can see in the example, we don't store denormalised documents for titles inside of genres, but a reference through a titles map indexed by id.
+I can show you some code that I've been working on for a proof of concept. Falcor server side is currently only available on Node.js but there are plans to porting it into another platforms.
```

Comment:
Falcor's server side is currently only available on Node.js but there are plans to port it to other platforms.
(include link for this information?)